planner/cascades/transformation_rules.go

```diff
@@ -1282,6 +1283,64 @@ func (r *PushTopNDownTiKVSingleGather) OnTransform(old *memo.ExprIter) (newExprs
 	return []*memo.GroupExpr{finalTopNExpr}, true, false, nil
 }
 
+// MergeAdjacentTopN merge adjacent TopN.
+type MergeAdjacentTopN struct {
+	baseRule
+}
+
+// NewRuleMergeAdjacentTopN creates a new Transformation MergeAdjacentTopN.
+// The pattern of this rule is `TopN->TopN->X`.
+func NewRuleMergeAdjacentTopN() Transformation {
+	rule := &MergeAdjacentTopN{}
+	rule.pattern = memo.BuildPattern(
+		memo.OperandTopN,
+		memo.EngineAll,
+		memo.NewPattern(memo.OperandTopN, memo.EngineAll),
+	)
+	return rule
+}
+
+// Match implements Transformation interface.
+func (r *MergeAdjacentTopN) Match(expr *memo.ExprIter) bool {
+	topN := expr.GetExpr().ExprNode.(*plannercore.LogicalTopN)
+	child := expr.Children[0].GetExpr().ExprNode.(*plannercore.LogicalTopN)
+	if len(topN.ByItems) != len(child.ByItems) {
+		return false
+	}
+	for i := 0; i < len(topN.ByItems); i++ {
+		if (*topN.ByItems[i]) != (*child.ByItems[i]) {
+			return false
+		}
```

Comment:
One last thing. This check is not correct. You should use
```
topN.ByItems[i].Expr.Equal(ctx, child.ByItems[i].Expr) && topN.ByItems[i].Desc == child.ByItems[i].Desc
```

BTW, you can just add a method `ByItem.Equal()`.